### PR TITLE
Handle exchange in Transaction Request flow

### PR DESCRIPTION
**Pull request by T-Dnzt**
Issue/Task Number: T446

# Overview

This PR adds the handling of exchanges to transaction requests.

# Changes

- Prevent exchange from happening on the client.
- Add exchange mechanism to transaction request.

### CODE REVIEW

File: apps/ewallet/lib/ewallet/web/v1/error_handler.ex [hunk 1]
```diff
@@ -215,6 +215,10 @@ defmodule EWallet.Web.V1.ErrorHandler do
     exchange_account_wallet_not_found: %{
       code: "exchange:account_wallet_not_found",
       description: "No wallet was found with the given exchange_wallet_address param."
+    },
+    exchange_not_allowed: %{
```
Comment (unnawut) [inline apps/ewallet/lib/ewallet/web/v1/error_handler.ex]: Can we rename to `exchange_client_not_allowed`? Thinking there might be other cases that we need to lock down.

File: apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_consumption_controller_test.exs [hunk 2]
```diff
@@ -525,5 +537,39 @@ defmodule EWalletAPI.V1.TransactionConsumptionControllerTest do
                }
              }
     end
+
+    test "fails to consume a comsumption that involves exchange",
```
Comment (unnawut) [inline apps/ewallet_api/test/ewallet_api/v1/controllers/transaction_consumption_controller_test.exs]: 👏 👏 👏 👏 👏

File: apps/ewallet_api/test/support/channel_case.ex [hunk 3]
```diff
@@ -13,7 +13,7 @@ defmodule EWalletAPI.ChannelCase do
   of the test unless the test case is marked as async.
   """
 
-  use ExUnit.CaseTemplate
+  use ExUnit.CaseTemplate, async: false
```
Comment (unnawut) [inline apps/ewallet_api/test/support/channel_case.ex]: T_T https://github.com/elixir-lang/elixir/blob/v1.6.6/lib/ex_unit/lib/ex_unit/case_template.ex#L32

File: apps/ewallet_db/lib/ewallet_db/transaction_consumption.ex [hunk 4]
```diff
@@ -189,6 +236,15 @@ defmodule EWalletDB.TransactionConsumption do
     end
   end
 
+  def get_final_amount(consumption) do
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/transaction_consumption.ex]: I think I need a `@spec` for this one. I'll probably forget why we need it a few weeks from now.

File: apps/ewallet_db/lib/ewallet_db/transaction_consumption.ex [hunk 5]
```diff
@@ -311,10 +367,31 @@ defmodule EWalletDB.TransactionConsumption do
...
+
+    {:ok, consumption} =
+      consumption
+      |> transaction_failure_changeset(data)
+      |> Repo.update()
+
+    consumption
+  end
+
+  # def fail(consumption, code) do
```
Comment (unnawut) [inline apps/ewallet_db/lib/ewallet_db/transaction_consumption.ex]: Comment?

File: apps/admin_api/test/admin_api/v1/controllers/admin_auth/transaction_consumption_controller_test.exs [hunk 6]
```diff
@@ -826,7 +826,10 @@ defmodule AdminAPI.V1.AdminAuth.TransactionConsumptionControllerTest do
                "version" => "1",
                "data" => %{
                  "address" => meta.account_wallet.address,
-                 "amount" => 100_000 * meta.token.subunit_to_unit,
+                 "amount" => nil,
+                 "estimated_consumption_amount" => 100_000 * meta.token.subunit_to_unit,
```
[the remark this answers is not in the document]
Comment (T-Dnzt) [reply apps/admin_api/test/admin_api/v1/controllers/admin_auth/transaction_consumption_controller_test.exs]: Added in `docs/guides/transaction_request_flow.md` :)

**Pull request merged**